### PR TITLE
Fix resource leak in tests and add load balancer

### PR DESCRIPTION
**Pull request by Kerram**
I merged the fix for tests with the introduction of load balancer, sorry :(

### CODE REVIEW

File: e2e_tests.py [hunk 1]
```diff
@@ -12,13 +13,20 @@
...
+    def stop_load_balancer(self):
+        deploy_utils.stop_fast_api_app(self.load_balancer_process)
 
     def setUp(self) -> None:
-        self.machine_manager_url = E2eTests.start_machine_manager()
+        self.machine_manager_process, self.machine_manager_url = deploy_utils.start_fast_api_app("machine_manager")
+        self.load_balancer_process, self.load_balancer_url = deploy_utils.start_fast_api_app("load_balancer")
+        load_balancer_api.set_machine_manager(self.load_balancer_url, self.machine_manager_url)
+
+    def tearDown(self) -> None:
```
Comment (Kerram) [inline e2e_tests.py:27]: I am sending SIGINT to our fastapi apps. If you create some new ones in tests, you need to manually stop them.
I will do it for linters in the next diff.

File: linter.py [hunk 2]
```diff
@@ -4,15 +4,26 @@
 
 linter_app = FastAPI()
 
+usage = 0
```
Comment (Kerram) [inline linter.py:7]: @tomg10 I am not sure, if this is the correct way to do it (but it works). I wanted to store a local variable in each linter instance, but I needed to use global usage in the validate_file function below.

I saw you didn't have to do so in the machine_manager.get_machines(). What is a difference?

File: linter_api.py [hunk 3]
```diff
@@ -7,3 +7,8 @@
 
 def validate(url, request: LinterRequest):
     return requests.post(f"{url}/validate", json=dataclasses.asdict(request)).json()
+
+
+def get_linter_usage(url):
```
Comment (Kerram) [inline linter_api.py:12]: It is "private" to be used only for testing. It is bad, but I didn't know a better way.

File: linter.py [hunk 4]
```diff
@@ -1,18 +1,29 @@
...
+
 
 @linter_app.get("/")
 def health_check() -> str:
-    return "ok"
+    return "ok linter"
```
Comment (mkmkl93) [inline linter.py:13]: Skąd ta zmiana?

File: linter.py [hunk 5]
```diff
@@ -1,18 +1,29 @@
...
 
 
 @linter_app.post("/validate")
 def validate_file(request: LinterRequest) -> LinterResponse:
-    return LinterResponse(result="ok", errors=[])
+    global usage
+
+    usage += 1
```
Comment (mkmkl93) [inline linter.py:20]: to jest chyba licznik ile razy odpowiedziało tak? to chyba bardziej adekwatną nazwą jest `responses_count` albo temu podobne

**Review by kirill-baryshnikov — COMMENTED** [review]
Please see the comments

File: e2e_tests.py [hunk 6]
```diff
@@ -13,20 +15,87 @@
 
 class E2eTests(unittest.TestCase):
     @staticmethod
-    def start_machine_manager():
-        _, machine_manager_url = deploy_utils.start_fast_api_app("machine_manager")
-        return machine_manager_url
+    def set_linter_debug_mode():
```
[the remark this answers is not in the document]
Comment (Kerram) [reply e2e_tests.py:18]: This is what we discussed on 13.01. We are using debug here to return additional data in the linter response to assert them in tests. It is not ideal, but a simplification we agreed to pursue.

File: linter.py [hunk 7]
```diff
@@ -1,18 +1,29 @@
...
 
 
 @linter_app.post("/validate")
 def validate_file(request: LinterRequest) -> LinterResponse:
-    return LinterResponse(result="ok", errors=[])
+    global responses_count
+
+    responses_count += 1
+
+    debug = []
```
[the remark this answers is not in the document]
Comment (Kerram) [reply linter.py:22]: Addressed in https://github.com/tomg10/mimuw_linters/pull/11.

File: load_balancer.py [hunk 8]
```diff
@@ -1,10 +1,39 @@
...
+def health_check() -> str:
+    return "ok balancer"
+
+
+@load_balancer_app.get("/set_machine_manager")
+def set_machine_manager(machine_manager: str = "") -> None:
+    global machine_manager_url
+
+    machine_manager_url = machine_manager
+    print(f"Machine manager set up on: {machine_manager_url}")
```
[the remark this answers is not in the document]
Comment (Kerram) [reply load_balancer.py:23]: Logs added in https://github.com/tomg10/mimuw_linters/pull/13.

File: load_balancer.py [hunk 9]
```diff
@@ -1,10 +1,39 @@
...
-@load_balancer_app.get("/validate")
+
+@load_balancer_app.get("/")
+def health_check() -> str:
+    return "ok balancer"
+
+
+@load_balancer_app.get("/set_machine_manager")
+def set_machine_manager(machine_manager: str = "") -> None:
+    global machine_manager_url
```
[the remark this answers is not in the document]
Comment (mkmkl93) [reply load_balancer.py:20]: Addressed in #11.

File: e2e_tests.py [hunk 10]
```diff
@@ -13,20 +15,87 @@
...
+        for linter in linters:
+            machine_manager_api.kill_linter_instance(self.machine_manager_url, linter.instance_id)
+
+        deploy_utils.stop_fast_api_app(self.load_balancer_process)
+        deploy_utils.stop_fast_api_app(self.machine_manager_process)
+
+        E2eTests.unset_linter_debug_mode()
+
+    def test_getting_linters(self):
         for i in range(10):
```
[the remark this answers is not in the document]
Comment (mkmkl93) [reply e2e_tests.py:43]: will be probablye solved with #18

File: e2e_tests.py [hunk 11]
```diff
@@ -13,20 +15,87 @@
...
+
+    def test_load_balancer_with_no_machines(self):
+        response: LinterResponse = LinterResponse.from_dict(
+            load_balancer_api.validate(self.load_balancer_url, LinterRequest(language="python", code="x=5")))
+
+        self.assertEqual("fail", response.result)
+        self.assertEqual(1, len(response.errors))
+        self.assertEqual("No linter machine available", response.errors[0])
+
+    def test_load_balancer_equal_split(self):
```
[the remark this answers is not in the document]
Comment (mkmkl93) [reply e2e_tests.py:91]: Point 3 and 4 are introduced in #4

File: deploy_utils.py [hunk 12]
```diff
@@ -10,3 +11,7 @@ def start_fast_api_app(app_name):
     process = subprocess.Popen(["uvicorn", f"{app_name}:{app_name}_app", f"--port={port}"])
```
[the remark this answers is not in the document]
Comment (mkmkl93) [reply deploy_utils.py:11]: Issue #18

File: e2e_tests.py [hunk 13]
```diff
@@ -13,20 +15,87 @@
...
+
+        machine_manager_api.kill_linter_instance(self.machine_manager_url, id1)
+
+        machines = machine_manager_api.get_machines(self.machine_manager_url)
+        self.assertEqual(1, len(machines))
+        self.assertEqual(id2, machines[0].instance_id)
+
+    def test_load_balancer_with_no_machines(self):
+        response: LinterResponse = LinterResponse.from_dict(
+            load_balancer_api.validate(self.load_balancer_url, LinterRequest(language="python", code="x=5")))
```
[the remark this answers is not in the document]
Comment (mkmkl93) [reply e2e_tests.py:85]: Issues #19 #20 #21

File: e2e_tests.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mkmkl93) [reply e2e_tests.py:91]: Issues #22 #23 #24 #25

File: load_balancer.py [hunk 14]
```diff
@@ -1,10 +1,39 @@
...
 def validate_file(request: LinterRequest) -> LinterResponse:
-    return LinterResponse(result="ok", errors=[])
+    global machine_number
+
+    machines = machine_manager_api.get_machines(machine_manager_url)
+
+    if machine_number >= len(machines):
+        machine_number = 0
+
+    if len(machines) == 0:
```
[the remark this answers is not in the document]
Comment (mkmkl93) [reply load_balancer.py:35]: Issue #24

File: deploy_utils.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (mkmkl93) [reply deploy_utils.py:11]: Resolved

File: e2e_tests.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mkmkl93) [reply e2e_tests.py:91]: #22 already implemented

**Pull request merged**